### PR TITLE
Fix SR-6060: marking +new unavailable when -init is unavailable

### PR DESCRIPTION
**Pull request by JaviSoto**
Currently a loop-hole exists by which one could end up invoking an unavailable `-init` initializer from Obj-C when it is unavailable by using `+new` (which itself calls `-init`)

Resolves [SR-5018](https://bugs.swift.org/browse/SR-5018).

### CODE REVIEW

File: lib/PrintAsObjC/PrintAsObjC.cpp [hunk 1]
```diff
@@ -612,6 +613,8 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
         // required access
         os << " SWIFT_UNAVAILABLE";
         skipAvailability = true;
+        // If -init is unavailable, then +new should be, too:
+        makeNewUnavailable = selectorPieces.size() == 1 && selectorPieces.front().str() == "init";
```
Comment (jrose-apple) [inline lib/PrintAsObjC/PrintAsObjC.cpp]: Teeeechnically this would catch `init:` as well. A better check would be `selector.getNumArgs()`.

**Review by jrose-apple — COMMENTED** [review]
This seems totally reasonable to me, with one comment. Thanks, Javi!

File: lib/PrintAsObjC/PrintAsObjC.cpp (hunk 1 above)
Comment (JaviSoto) [reply lib/PrintAsObjC/PrintAsObjC.cpp]: So `selectorPieces.size() == 1 && selector.getNumArgs() == 0 && selectorPieces.front().str() == "init";`?

Comment (jrose-apple) [reply lib/PrintAsObjC/PrintAsObjC.cpp]: `selector.getNumArgs()` should be sufficient. There's no way to have a multi-piece selector with zero arguments, and there are no zero-piece selectors.

Comment (JaviSoto) [reply lib/PrintAsObjC/PrintAsObjC.cpp]: Ah duh of course. Ok fixed, thanks!
I have only run the tests for `PrintAsObjC` locally, I assume you have to tell CI to run for this?

**Review by jrose-apple — APPROVED** [review]

File: lib/PrintAsObjC/PrintAsObjC.cpp (hunk 1 above)
Comment (jrose-apple) [reply lib/PrintAsObjC/PrintAsObjC.cpp]: Yep, I got it.

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
And though I really hope this is unnecessary…

@swift-ci Please test source compatibility

**Comment by jrose-apple** [comment]
Shows how good my code review is. `selector`, not `selectorPieces`.

**Comment by JaviSoto** [comment]
Nah that’s totally my bad for not running the tests locally again. Ok should be fixed now!

As for the source compatibility: I should’ve pointed out that this is a potentially source breaking change, because you could’ve been calling +new on a Swift class where init wasn’t available before, and now that’d be a build error. However, I think it’s easy to argue that that’s a good thing, because that was potentially a bug before.

**Comment by jrose-apple** [comment]
Yep, the fact that calling +new would have crashed at run time is why I'm not as worried about this. Running the source compat suite just checks if we *know* someone happens to be doing it anyway, in which case we could consider downgrading it to a warning in Swift 4 mode or something.

@swift-ci Please test

**Comment by jrose-apple** [comment]
@swift-ci Please test source compatibility

**Comment by jrose-apple** [comment]
@swift-ci Please test source compatibility

**Comment by JaviSoto** [comment]
Is this good to go?

**Comment by jrose-apple** [comment]
Yep, sorry to let it sit on Friday. Thanks, Javi!

**Comment by JaviSoto** [comment]
No worries at all! my pleasure!

**Pull request merged**